### PR TITLE
doc(davide) GAIL + D4PG + TD3BC + SAC: doc and benchmark

### PR DESCRIPTION
**Pull request by davide97l**
Added or enhanced documentation and benchmark of the following algorithms:

- SAC
- GAIL
- D4PG
- TD3BC

### CODE REVIEW

File: source/hands_on/d4pg.rst [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+----------------------------
+
+- `D4PG Pytorch`_
+
+.. _`D4PG Pytorch`: https://github.com/schatty/d4pg-pytorch
+
+
+References
+-----------
+Gabriel Barth-Maron, Matthew W. Hoffman, David Budden, Will Dabney, Dan Horgan, Dhruva TB, Alistair Muldal, Nicolas Heess, Timothy Lillicrap: “Distributed Distributional Deterministic Policy Gradients”, 2018; [https://arxiv.org/abs/1804.08617v1 arXiv:1804.08617v1].
```
Comment (Weiyuhong-1998) [inline source/hands_on/d4pg.rst:110]: 用超链接的形式？

File: source/hands_on/d4pg.rst [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   :scale: 75%
+   :align: center
+
+
+Implementations
+----------------
+The default config is defined as follows:
+
+.. autoclass:: ding.policy.d4pg.D4PGPolicy
+   :noindex:
```
Comment (Weiyuhong-1998) [inline source/hands_on/d4pg.rst:61]: add more comments in d4pg policy config?

File: source/hands_on/gail.rst [hunk 3]
```diff
@@ -80,6 +82,40 @@ Extensions
...
+|(BipedalWalker-v3)   |                 |                                                     |config/bipedalwalker_sac_ |                      |
+|                     |                 |                                                     |gail_config.py>`_         |                      |
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
+|                     |                 |                                                     |`config_link_h <https://  |                      |
+|                     |                 |                                                     |github.com/opendilab/     |                      |
+|Hopper               |                 |                                                     |DI-engine/tree/main/dizoo/| SAC                  |
+|                     |  3500           |.. image:: images/benchmark/hopper_gail.png          |mujoco/config/hopper_sac_ |                      |
+|(Hopper-v2)          |                 |                                                     |gail_default_config.py>`_ |                      |
```
Comment (puyuan1996) [inline source/hands_on/gail.rst:116]: Hopper-v2 or Hopper-v3?

Comment (davide97l) [reply source/hands_on/gail.rst:116]: You are right, is v3, thank you for pointing it out.

**Comment by davide97l** [comment]
> GAIL is a new general framework for directly extracting policy from data, as if it were obtained by reinforcement learning following inverse reinforcement learning.

Added in the introduction

File: source/hands_on/d4pg.rst [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+1. D4PG is only used for environments with **continuous action spaces**.(i.e. MuJoCo)
+
+2. D4PG is an **off-policy** algorithm.
+
+3. D4PG uses a **distributional** critic.
+
+4. D4PG is a **model-free** and **actor-critic** RL algorithm, which optimizes actor network and critic network, respectively.
+
+5. Usually, D4PG use **Ornstein-Uhlenbeck process** or **Gaussian process** (default in our implementation) for exploration.
+
```
Comment (ccicc1) [inline source/hands_on/d4pg.rst:24]: D4PG uses

File: source/hands_on/d4pg.rst [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Key Equations or Key Graphs
+---------------------------
+The D4PG algorithm maintains a distributional critic :math:`Z_\pi(x, a)` and the loss used to learn the distribution parameters is defined as :math:`L(\pi) = \mathbb{E}[d(\mathcal{T}_{\mu_\theta}, Z_{pi'}(s, a), Z_\pi(s, a)]`, where :math:`\mathcal{T}_{\mu_\theta}` si the Bellman operator. The implemented distribution is usually a Categorical distribution.
```
Comment (ccicc1) [inline source/hands_on/d4pg.rst:27]: where :math:`\mathcal{T}_{\mu_\theta}` si the Bellman operator
si 好像打错了

File: source/hands_on/td3_bc.rst [hunk 6]
```diff
@@ -0,0 +1,104 @@
+TD3BC
+^^^^^^^
+
+Overview
+---------
+
+TD3BC, proposed in the 2021 paper `A Minimalist Approach to Offline Reinforcement Learning <https://arxiv.org/abs/2106.06860>`_,
+is a simple approach to offline RL where only two changes are made to TD3: a weighted behavior cloning loss is added to the policy update and the states are normalized.
+Unlike competing methods there are no changes to architecture or underlying hyperparameters.
+The resulting algorithm is a simple to implement and tune baseline, while more than halving the overall run time by removing the additional computational overhead of previous methods.
```
Comment (ccicc1) [inline source/hands_on/td3_bc.rst:10]: The resulting algorithm is a sample?

File: source/hands_on/d4pg.rst (hunk 2 above)
Comment (davide97l) [reply source/hands_on/d4pg.rst:61]: I think the 'config' section is not shown because of a format problem in di-engine. I will investigate it and change accordingly.

File: source/hands_on/td3_bc.rst (hunk 6 above)
Comment (davide97l) [reply source/hands_on/td3_bc.rst:10]: this sentence is correct and has been used also by the authors of the paper: is 'simple to implement' and 'tune' baseline. Basically the baseline benefits of both properties.

File: source/hands_on/d4pg.rst (hunk 4 above)
Comment (davide97l) [reply source/hands_on/d4pg.rst:24]: corrected, thank you!

File: source/hands_on/d4pg.rst (hunk 5 above)
Comment (davide97l) [reply source/hands_on/d4pg.rst:27]: yes, there is a typo here. good find

File: source/hands_on/td3_bc.rst (hunk 6 above)
Comment (davide97l) [reply source/hands_on/td3_bc.rst:10]: I will rewrite like this to make it more clear: is a simple baseline that is easy to implement and to tune

File: source/hands_on/sac.rst [hunk 7]
```diff
@@ -270,7 +270,35 @@ Auto alpha strategy
             self._alpha_optim.step()
             self._alpha = self._log_alpha.detach().exp()
 
-The Benchmark result of SAC implemented in DI-engine is shown in `Benchmark <../feature/algorithm_overview.html>`_
```
Comment (Will-Nie) [inline source/hands_on/sac.rst:273]: Comments:
1. In the Overview section, can we describe more about sac? For instance: What is sac, which categorisation does sac belong to? Why this algorithm is proposed and what's the final performance it achieves. A small trick would be to cite some sentences from the paper and rephrase them.

File: source/hands_on/sac.rst [hunk 8]
```diff
@@ -270,7 +270,35 @@ Auto alpha strategy
...
+|                     |                 |                                                     |                          |                      |
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
+|                     |                 |                                                     |`config_link_ho <https:// |                      |
+|                     |                 |                                                     |github.com/opendilab/     | Baselines(2325)      |
+|Hopper               |                 |                                                     |DI-engine/blob/main/dizoo/|                      |
+|                     |  3653           |.. image:: images/benchmark/hopper_sac.png           |mujoco/config/hopper_sac_ | Tianshou(3542)       |
+|(Hopper-v2)          |                 |                                                     |default_config.py>`_      |                      |
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
```
Comment (Will-Nie) [inline source/hands_on/sac.rst:300]: shall we add Spinning-up as well for comparison. Also, it might be better to replace Baselines by Sb3

File: source/hands_on/gail.rst [hunk 9]
```diff
@@ -80,6 +85,40 @@ Extensions
...
+Benchmark
+-----------
+
+
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
+| environment         |best mean reward | evaluation results                                  | config link              | expert               |
++=====================+=================+=====================================================+==========================+======================+
+|                     |                 |                                                     |`config_link_l <https://  |                      |
+|                     |                 |                                                     |github.com/opendilab/     |                      |
+|                     |                 |                                                     |DI-engine/tree/main/dizoo/| DQN                  |
```
Comment (norman26625) [inline source/hands_on/gail.rst:103]: line 104 and line 106 spelling of Lunarlander is different

File: source/hands_on/gail.rst [hunk 10]
```diff
@@ -80,6 +85,40 @@ Extensions
...
+|                     |                 |                                                     |lunarlander_dqn_gail_     |                      |
+|(LunarLander-v2)     |                 |                                                     |config.py>`_              |                      |
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
+|                     |                 |                                                     |`config_link_b <https://  |                      |
+|                     |                 |                                                     |github.com/opendilab/     |                      |
+|Bipedalwalker        |                 |                                                     |DI-engine/tree/main/dizoo/| SAC                  |
+|                     |  300            |.. image:: images/benchmark/bipedalwalker_gail.png   |box2d/bipedalwalker/      |                      |
+|(BipedalWalker-v3)   |                 |                                                     |config/bipedalwalker_sac_ |                      |
```
Comment (norman26625) [inline source/hands_on/gail.rst:112]: line 112 and line 110 spelling of BipedalWalker is different

File: source/hands_on/gail.rst [hunk 11]
```diff
@@ -80,6 +85,40 @@ Extensions
...
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
+|                     |                 |                                                     |`config_link_h <https://  |                      |
+|                     |                 |                                                     |github.com/opendilab/     |                      |
+|Hopper               |                 |                                                     |DI-engine/tree/main/dizoo/| SAC                  |
+|                     |  3500           |.. image:: images/benchmark/hopper_gail.png          |mujoco/config/hopper_sac_ |                      |
+|(Hopper-v3)          |                 |                                                     |gail_default_config.py>`_ |                      |
++---------------------+-----------------+-----------------------------------------------------+--------------------------+----------------------+
+
 Reference
 ---------
```
Comment (norman26625) [inline source/hands_on/gail.rst:124]: reference has a standard format, if you can't find it, please contact me (hs)
e.g.:https://github.com/opendilab/DI-engine-docs/blob/main/source/hands_on/r2d3_zh.rst

File: source/hands_on/gail.rst (hunk 10 above)
Comment (davide97l) [reply source/hands_on/gail.rst:112]: changed both to double capital letter version

File: source/hands_on/sac.rst (hunk 8 above)
Comment (davide97l) [reply source/hands_on/sac.rst:300]: added

File: source/hands_on/sac.rst (hunk 7 above)
Comment (davide97l) [reply source/hands_on/sac.rst:273]: 1. 2. 3. 4. 5. Solved
6. To check later

File: source/hands_on/td3_bc.rst [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+Key Equations or Key Graphs
+---------------------------
+TD3BC simply consists to add a behavior cloning term to TD3 in order to regularize the policy:
+
+.. math::
+    \begin{aligned}
+    \pi = \arg\max_{\pi} \mathbb{E}_{(s, a) \sim D} [ \lambda Q(s, \pi(s)) - (\pi(s)-a)^2 ]
+    \end{aligned}
+
+Additionally, all the states in the dataset are normalized, such that they have mean 0 and standard deviation 1.
```
Comment (PaParaZz1) [inline source/hands_on/td3_bc.rst:27]: state normalization is applied in each mini-batch, current statements may lead to misunderstanding, such as norm in the whole dataset.

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: source/hands_on/td3_bc.rst [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+~~~~~~~~~~~~~~~~~
+Here we provide examples of `td3_bc` model as default model for `TD3BC`.
+
+.. autoclass:: ding.model.template.qac.QAC
+    :members: forward, compute_actor, compute_critic
+    :noindex:
+
+
+Benchmark
+-----------
```
[the remark this answers is not in the document]
Comment (davide97l) [reply source/hands_on/td3_bc.rst:47]: added as a note after the benchmark table

File: source/hands_on/td3_bc.rst [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+
+- `Official implementation`_
+- d3rlpy_
+
+.. _`Official implementation`: https://github.com/sfujim/TD3_BC
+.. _d3rlpy: https://github.com/takuseno/d3rlpy
+
+References
+-----------
+Scott Fujimoto, Shixiang Shane Gu: “A Minimalist Approach to Offline Reinforcement Learning”, 2021; [https://arxiv.org/abs/2106.06860 arXiv:2106.06860].
```
Comment (Will-Nie) [inline source/hands_on/td3_bc.rst:112]: swap `references` and `Other Public Implementations`

Comment (davide97l) [reply source/hands_on/td3_bc.rst:112]: closed in latest commit

File: source/hands_on/td3_bc.rst [hunk 15]
```diff
@@ -98,15 +98,15 @@ Benchmark
...
 - `Official implementation`_
 - d3rlpy_
 
 .. _`Official implementation`: https://github.com/sfujim/TD3_BC
-.. _d3rlpy: https://github.com/takuseno/d3rlpy
-
-References
------------
-Scott Fujimoto, Shixiang Shane Gu: “A Minimalist Approach to Offline Reinforcement Learning”, 2021; [https://arxiv.org/abs/2106.06860 arXiv:2106.06860].
+.. _d3rlpy: https://github.com/takuseno/d3rlpy
```
Comment (ccicc1) [inline source/hands_on/td3_bc.rst:112]: 此处格式需要与以往文档统一

**Comment by davide97l** [comment]
@ccicc1 Thank you for pointing out, I will check and correct these spacing problems. However, now this PR has been divided into single PRs, one for each algorithm. For example, td3bc is now in: https://github.com/opendilab/DI-engine-docs/pull/100.

**Comment by davide97l** [comment]
> 这段不该有的空格去一下，比如传统的Q- learning/使用深度神经网络来估计Q值

Could you specify in which line?
